lib/absinthe/phoenix/controller.ex

```diff
@@ -0,0 +1,320 @@
+defmodule Absinthe.Phoenix.Controller do
+  @moduledoc """
+  Supports use of GraphQL documents inside Phoenix controllers.
+
+  ## Example
+
+  First, `use Absinthe.Phoenix.Controller`, passing your `schema`:
+
+  ```elixir
+  defmodule MyAppWeb.UserController do
+    use MyAppWeb, :controller
+    use Absinthe.Phoenix.Controller, schema: MyAppWeb.Schema
+
+    # ... actions
+
+  end
+  ```
+
+  For each action you want Absinthe to process, provide a GraphQL document using
+  the `@graphql` module attribute (before the action):
+
+  ```
+  @graphql \"""
+    query ($filter: UserFilter) {
+      users(filter: $filter, limit: 10)
+    }
+  \"""
+  def index(conn, %{data: data}) do
+    render conn, "index.html", data
+  end
+  ```
+
+  The params for the action will be intercepted by the
+  `Absinthe.Phoenix.Controller.Action` plug, and used as variables for
+  the GraphQL document you've specified.
+
+  For instance, given a definition for a `:user_filter` input object
+  type like this:
+
+  ```
+  input_object :user_filter do
+    field :name_matches, :string
+    field :age_above, :integer
+    field :age_below, :integer
+  end
+  ```
+
+  And a query that looks like this (assuming you have the normal
+  `Plug.Parsers` configuration for param parsing):
+
+  ```
+  ?filter[name_matches]=joe&filter[age_above]=42
+  ```
+
+  Then Absinthe will receive variable definitions of:
+
+  ```
+  %{"filter" => %{"name_matches" => "joe", "age_above" => 42}}
+  ```
+
+  (For how the string `"42"` was converted into `42`, see `cast_param/3`).
+
+  The params on the `conn` will then be replaced by the result of the
+  execution by Absinthe. The action function can then match against
+  that result to respond correctly to the user:
+
+  It's up to you to handle the three possible results:
+
+  - When there's `:data` but no `:errors`, everything went perfectly.
+  - When there's `:errors` but no `:data`, a validation error occurred and the document could not be
+    executed.
+  - When there's `:data` and `:errors`, partial data is available but some fields reported errors
+    during execution.
+
+  Notice the keys are atoms, not strings as in normal Phoenix action invocations.
+
+  ## Differences with the GraphQL Specification
+
+  There are some important differences between GraphQL documents as
+  processed in an HTTP API and the GraphQL documents that this module
+  supports.
+
+  In an effort to make use of GraphQL ergonomic in Phoenix controllers
+  and views, Absinthe supports some slight structural modifications to
+  the GraphQL documents provided using the `@graphql` module attribute
+  in controller modules.
+
+  In a way, you can think of these changes as a specialized GraphQL
+  dialect. The following are the differences you need to keep in mind.
+
+  ### Objects can be leaf nodes
+
+  Let's look at the `users` example mentioned before:
+
+  ```
+  @graphql \"""
+    query ($filter: UserFilter) {
+      users(filter: $filter, limit: 10)
```

Comment:
@bruce how does this work with associations? IE what if I do 
```
{
  users(filter: $filter, limit: 10) {
    organization
  }
}
```

Does this only show the organization now? Or does it always return any non null value present?

Reply:
As we discussed earlier today: currently if you add a selection for a field (eg, as `users` has here), you don't get the full structs there. If we modify this to use explicit directives (or a similar mechanism), perhaps we can support both. Let's see after we get more feedback.